### ISSUE TITLE
Unclosed GZIPOutputStream cause mass memory occupation while stress testing

### ISSUE THREAD

**Issue by v-gu** [gwhalin/Memcached-Java-Client#2]
Recently our developer was evaluating Memcached as an alternative solution on site specific problems. When we started stress test on Memcached with compress enabled, something went wrong.
We call the set(...) method of MemcachedClient within a loop with a large base, like one million times per batch; and we limited the Java heap size up to 2gb, the stack size up to 1024k. After the test began, the Resident Memory of this Java process inflated dramaticly in liner pattern, without the acknowledgement of heap limit. Our test server has 16GB physical memory, but in a short while, within 40 seconds approximately, this test Java process had taken up all the physical memory and throws OOM with the exception shows the zlib was just trying to allocate memory. 
Then we've profiled the test process but found nothing abnormal, so we dumped the heap, and it was okay, only occupies 123mb, and the stack was also okay; as the GZIPOurputStream utilizing JNI call, we believe this problem was caused by over-inflation of native heap, the memory malloced by JNI function. 
So we examined the GZIPOutputStream and found it extends DeflaterOutputStream and it use Deflater to compress stream, according to Java-doc of Deflater, the caller should invoke the end() method of Deflater object when the compressor is no longer being used or it will also be called automatically by the finalize() method. So here is the problem, now the MemcachedClient code does not explicitly invoke the end() method(which is indirectly invoked by the close() method of GZIPOutputStream), then the Java VM may not realize the mass memory occupied by JNI portion, and the Deflator object may not be GCed at first priority(in previously dumped heap, the Deflator class has approximately 6 thousand instances), at this point, the native heap memory was out of control.
To sum up, we insert a simple line after line 800:
    gos.finish();
    gos.close();
then the memory problem definitely gone, all seems good afterward, the Java process Resident Memory had never gone too far, just around 800mb.

So should we add this line to the trunk code?

**Comment by gwhalin** [comment]
Fixed in trunk.  Thanks for the catch.
